### PR TITLE
 [FLINK-7420] [avro] Move all Avro code to flink-avro (followup)

### PR DESCRIPTION
**Pull request by StephanEwen**
## What is the purpose of the change

This is an extension of #4931 which adds some cleanups and improvements.
Most notably, it makes the new `flink-avro` module independent of runtime dependencies and Scala versions.

## Brief change log

  - Move the `SerializationSchema` classes to `flink-core` to make them accessible from API projects without introducing runtime dependencies. This is done in a non-API-breaking way.
  - Make `flink-avro` module independent of `flink-streaming-java`, `flink-runtime` and hence independent of Scala versions.
  - Add various cleanups and fixes
  - Adds a test that validates that a KryoSerializer from Flink 1.3 (which has implicit Avro dependency classes) is still deserializable with the state serializer utils (that implement the serializer and state evolution).

## Verifying this change

  - Most of the changes are covered by existing tests
  - Adds a test that validates that a KryoSerializer from Flink 1.3 (which has implicit Avro dependency classes) is still deserializable with the state serializer utils (that implement the serializer and state evolution).

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (**yes** / no): **This changes the `flink-avro` module to be Scala independent.

  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**yes** / no): **Moves the SerializationSchema classes, but leaves old classes in place (extending new classes) to preserve the API compatibility**.

  - The serializers: (yes / **no** / don't know) 
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)

  - 
## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by aljoscha** [comment]
The follow-up changes look good! (it's very easy to review since it's clearly separated. 😃)

If you reviewed Timo's and my changes I would say this is good to go.

**Comment by StephanEwen** [comment]
Lets see if Travis agrees. That guy has opinions...

**Comment by aljoscha** [comment]
Personally, I like this Jenkins fellow more... 😉

File: flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/KryoRegistrationSerializerConfigSnapshot.java [hunk 1]
```diff
@@ -217,7 +217,7 @@ public void read(DataInputView in) throws IOException {
 	/**
 	 * Placeholder dummy for a previously registered class that can no longer be found in classpath on restore.
 	 */
-	public static class DummyRegisteredClass {}
+	public static class DummyRegisteredClass implements Serializable {}
```
Comment (tillrohrmann) [inline flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/KryoRegistrationSerializerConfigSnapshot.java]: `serialVersionUID` is missing.

**Comment by tillrohrmann** [comment]
Great fix for our Avro dependency @StephanEwen, @aljoscha and @twalthr. +1 for merging.

**Pull request merged**